File: backend/phonics.py
"""Phonics data access layer — pure database pass-through, no runtime analysis.

All phonics annotation is pre-generated offline. This module only reads
stored data and returns it directly. Manual overrides take priority.

v2 schema (segments nested inside syllables):
{
    "word": "knowledge",
    "ipa": "/ˈnɒlɪdʒ/",
    "arpabet": "N AA1 L AH0 JH",
    "syllables": [{
        "text": "know",
        "stress": 1,
        "segments": [
            {"text": "k", "silent": true,  "rule": "silent-k"},
            {"text": "n", "silent": false, "rule": "consonant-n"},
            {"text": "o", "silent": false, "rule": "vowel-o"},
            {"text": "w", "silent": true,  "rule": "silent-w"}
        ]
    }, {
        "text": "ledge",
        "stress": 0,
        "segments": [
            {"text": "l",  "silent": false, "rule": "consonant-l"},
            {"text": "e",  "silent": false, "rule": "schwa"},
            {"text": "d",  "silent": true,  "rule": "silent-d-in-dge"},
            {"text": "ge", "silent": false, "rule": "dge"}
        ]
    }]
}
"""
import json
import os


# ---------------------------------------------------------------------------
# Manual overrides — teacher-reviewed data that always takes priority
# ---------------------------------------------------------------------------
_manual_overrides = None
_manual_overrides_path = os.path.join(
    os.path.dirname(__file__), '..', 'data', 'manual_overrides.json'
)


def _load_manual_overrides():
    """Load manual_overrides.json once per process."""
    global _manual_overrides
    if _manual_overrides is not None:
        return _manual_overrides
    try:
        with open(_manual_overrides_path, 'r', encoding='utf-8') as f:
            _manual_overrides = json.load(f)
    except (FileNotFoundError, json.JSONDecodeError):
        _manual_overrides = {}
    return _manual_overrides


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def get_phonics_data(word_model_instance):
    """Return phonics_data from a Word model instance, or None.

    Manual overrides in data/manual_overrides.json always take priority
    over database-stored data.
    """
    word_lower = word_model_instance.english.lower().strip()

    # 1. Check manual overrides first (always wins)
    overrides = _load_manual_overrides()
    if word_lower in overrides:
        entry = overrides[word_lower]
        if entry.get('reviewed') and 'phonics_data' in entry:
            return entry['phonics_data']

    # 2. Check database
    if word_model_instance.phonics_data is None:
        return None

    return word_model_instance.phonics_data


def adapt_phonics_for_frontend(phonics_data):
    """Pass phonics_data directly to frontend — no transformation needed.

    v2 format: segments are nested inside syllables. Frontend walks
    syllables[].segments[] directly.
    """
    return phonics_data

File: flutter_app/lib/models/phonics.dart
/// Phonics DTOs — mapped from GET /api/words/phonics.
///
/// Backend v2 schema: segments are nested inside syllables.
library;

class PhonicsData {
  const PhonicsData({
    required this.word,
    this.ipa,
    this.arpabet,
    required this.syllables,
  });

  final String word;
  final String? ipa;
  final String? arpabet;
  final List<PhonicsSyllable> syllables;

  factory PhonicsData.fromJson(Map<String, dynamic> json) {
    final rawSyllables = json['syllables'] as List<dynamic>? ?? [];
    return PhonicsData(
      word: json['word'] as String,
      ipa: json['ipa'] as String?,
      arpabet: json['arpabet'] as String?,
      syllables: rawSyllables
          .map((s) => PhonicsSyllable.fromJson(s as Map<String, dynamic>))
          .toList(growable: false),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'word': word,
      'ipa': ipa,
      'arpabet': arpabet,
      'syllables': syllables.map((s) => s.toJson()).toList(),
    };
  }
}

class PhonicsSyllable {
  const PhonicsSyllable({
    required this.text,
    required this.stress,
    required this.segments,
  });

  final String text;
  final int stress;
  final List<PhonicsSegment> segments;

  factory PhonicsSyllable.fromJson(Map<String, dynamic> json) {
    final rawSegments = json['segments'] as List<dynamic>? ?? [];
    return PhonicsSyllable(
      text: json['text'] as String,
      stress: json['stress'] as int,
      segments: rawSegments
          .map((s) => PhonicsSegment.fromJson(s as Map<String, dynamic>))
          .toList(growable: false),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'text': text,
      'stress': stress,
      'segments': segments.map((s) => s.toJson()).toList(),
    };
  }
}

class PhonicsSegment {
  const PhonicsSegment({
    required this.text,
    required this.silent,
    required this.rule,
  });

  final String text;
  final bool silent;
  final String rule;

  factory PhonicsSegment.fromJson(Map<String, dynamic> json) {
    return PhonicsSegment(
      text: json['text'] as String,
      silent: json['silent'] as bool,
      rule: json['rule'] as String,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'text': text,
      'silent': silent,
      'rule': rule,
    };
  }
}
